PID notches and D feed-foward

andyp1per opened the pull request:
This is a PR for discussion and experimentation only. We have discussed how notch filtering of the target and error terms might be useful for things like frame resonance and how allowing target D feedforward might bring some of the benefits in control that betaflight see. This change lumps them all together so that they can be tried.

IamPete1 commented:
https://github.com/ArduPilot/ardupilot/pull/23432 is a similar idea of notches on the target to prevent excitation of frame resonances. When reading up for that we found notches were not commonly used for that type of input shaping there are other methods that can work better. I think its fine to see if it works with notches but I think exporting some of the standard methods would be worth while as it might result in something that works better and needs less configuration. 

I have heard of derivative on measurement before, but never derivative on target, is there a betaflight name for it?

andyp1per commented:
It turns out that what betaflight calls "feed forward" is actually "derivative feedforward"

IamPete1 commented:
My understanding of betaflight is that acro is direct rate control, so "feed forward" is more of a pilot feel parameter than a control system one? Sort of like stick expo.

andyp1per commented:
> My understanding of betaflight is that acro is direct rate control, so "feed forward" is more of a pilot feel parameter than a control system one? Sort of like stick expo.

You should check with @lthall - I am just following his lead here

andyp1per commented:
You can see the notch on E at work here: 
![image](https://github.com/ArduPilot/ardupilot/assets/2893260/bb2c943a-25d9-473c-b217-e63c5c6cfb65)

andyp1per commented:
D feed forward is a game changer! In this flip you can see that it is doing all of the heavy lifting with P&D merely cleaning up the remaining noise:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/3520e7b5-f01f-45f5-b772-a5f4b0a1cb33)

Contrast without where P&D are having to do all the work:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/08933b8f-96ae-4e10-ae07-f290dedeea92)

This results in much better position holding in aggressive maneuvers:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/a9103acf-22a0-4c60-af6c-eb4a22af0295)

vs without:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/d1e976e5-ef9b-49d5-889c-c0b74eb218b9)

andyp1per commented:
Here is a real copter with massive frame resonance on roll:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/986ee213-f721-4383-b8be-e688c37fe33d)

and putting a single notch on roll in the PID completely eliminates this:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/f0d372ad-247e-45ec-9968-c849e8c391fd)

so appears to be working well. @bnsgeyer would love your feedback/testing on this

IamPete1 reviewed (dismissed):
I think these two features should be separate PR's. 

Would also be good to see some with and without logs.

tridge reviewed (changes requested):
I think this should be re-worked as a lua binding that allows the addition of a notch onto a PID. It can create parameters in the same place that this PR does (scripts can add parameters into existing parameter trees) but it will mean the parameters appear only when used and will allow us to rapidly iterate the development of this feature

bnsgeyer commented:
@andyp1per and @lthall Finally had the chance to test the PID notch filter feature.  This is a great feature that will provide a lot of flexibility to address frame resonances and for heli users, lightly damped rotor modes.  I used this new feature to address the lightly damped pitch mode of my T-Rex 700.  After doing an initial tune of the Pitch rate P and Rate D gains, the pitch axis has a lightly damped mode at approximately 4 hz.  With the pitch PID notch filter, I was able to shape the frequency response of pitch axis to reduce the effects of the lightly damped mode.  The plot shows using the PID notch filter to filter the target, error and both target and error.  In the case of heli's, it seems that it is best to filter the target signal.  The notch filter was set to a center freq of 4.4 hz, bandwidth of 2.2hz and depth of 5 db.
![image](https://github.com/ArduPilot/ardupilot/assets/9907906/eec2776d-d263-4262-8d0e-be8cf7a59da8)

bnsgeyer commented:
> I think this should be re-worked as a lua binding that allows the addition of a notch onto a PID. It can create parameters in the same place that this PR does (scripts can add parameters into existing parameter trees) but it will mean the parameters appear only when used and will allow us to rapidly iterate the development of this feature

@tridge  I'm not sure what there is to do to further develop this feature.  I  don't see why we shouldn't just merge it the way it is.

lthall commented:
I agree with Bill here.

andyp1per commented:
When we reviewed last week the push back revolved around two items:

1. We don't know what we need so need a lot of flexibility
2. There are too many parameters.

Fundamentally I do not agree with (1) - the proposal here has come out of a lot of discussion and experience so the problems are clear and the solution is clear as well. I believe we do know what we need and am not keen on changing the implementation to reflect some future that is unknown.

For point (2) I think there are a few options:
1. Introduce a _NEN notch enable parameter that hides the notch parameters unless set. Due to our level of nesting this could only be in-line so we probably want to have an index that allows for expansion on both sides of the divide - e..g. 32
2. Introduce a _ADV advanced parameter enablement parameter and hide both the notch parameters and things like PDMAX and SMAX behind it. For existing parameters we would need to do some migration. Or could just go for things after SMAX.
3. I do question the benefit of these two - who are we hiding these from? People won't see the new parameters in Mission Planner extended tuning page - it would only be in full parameter list. So I'm not convinced there is a big user benefit here.
4. Limit the scope of the new parameters by introducing a new AC_PID_Advanced that is only used by Copter and Heli - that would mean other vehicles could take on the new stuff as they desired.

andyp1per commented:
Added to devcall at @lthall 's request

IamPete1 commented:
Thought I would comment here rather than wait for the call so @andyp1per gets a chance to respond. 

Ultimately I think this PR can be boiled down to three features, one of which I think is beneficial in theory and thanks to @bnsgeyer 's testing has been proven to be beneficial in practice too. 

1. PID Target notch filter. 
This is the bit I think we should merge. There is a clear benefit in dealing with resonances. A notch on the target prevents the controller exciting the resonance. 

  I would add here that it would have been nice to have a look at some other methods of input shaping, my understanding is that there are other methods that both work better and require less configuration, notch filters are not typically used. However, notches are better than nothing. 

2. PID Error notch filter. 
I don't think this makes sense, notching the error prevents the controller from reacting at the given frequency at all. If you set your notch at xHz and your vehicle has a xHz oscillation the controller will never see the oscillation, it will not react. Your turning off the PID controller at xHz. This is why you filter the target, you cannot excite the resonance but you can react to it. 

3. D feed forward. 
I'm yet to be convinced that this is beneficial. I don't think the "betaflight has it" argument holds up since it's acro mode does not work the same was as ours (please correct me if I'm wrong). The angle control is better in the 1 test that is shown here, but the rate control look the same. I would like to see some more testing over a wider range of vehicles (possibly that has been done already, but if it's not documented on the PR it doesn't count ;) ). 


The TLDR is that I don't think we should merge this PR as it is. If the PID target filtering was broken out into a new PR I would support merging that feature.

bnsgeyer commented:
> I would add here that it would have been nice to have a look at some other methods of input shaping, my understanding is that there are other methods that both work better and require less configuration, notch filters are not typically used. However, notches are better than nothing.

@IamPete1 I would be interested to hear more about these.  I am not familiar with what else could be used to do this but that isn't saying much ;)  

> PID Error notch filter.
I don't think this makes sense, notching the error prevents the controller from reacting at the given frequency at all. If you set your notch at xHz and your vehicle has a xHz oscillation the controller will never see the oscillation, it will not react. Your turning off the PID controller at xHz. This is why you filter the target, you cannot excite the resonance but you can react to it.

I disagree with this.  It makes perfect sense when you are dealing with feedback instabilities or servoaeroelastic instabilities.  I didn't talk about it in the case I showed above but the only reason I didn't have to use the error notch was that I tuned the rate controller so that it didn't drive feedback instability.  I help a company deal with a bad tail wag due to the feedback loops coupling with the tailboom/gear when it was landed.  I used the 2nd harmonic notch as a static notch to prevent the tail wag.  Luckily the frequency was above the control  frequencies.  Having the notch in the PID would allow for addressing these types of feedback instabilities per axis and not penalizing other axes especially for issues near control frequencies.

The reason heli's need the notch on the target is due to the use of the FF gain in the rate controller as the primary control method which allows the pilot or autopilot to excite the lightly damped rotor mode.  This could also be due in part to the correction added by the attitude controller through angle feedback.

> D feed forward.
I'm yet to be convinced that this is beneficial. I don't think the "betaflight has it" argument holds up since it's acro mode does not work the same was as ours (please correct me if I'm wrong). The angle control is better in the 1 test that is shown here, but the rate control look the same. I would like to see some more testing over a wider range of vehicles (possibly that has been done already, but if it's not documented on the PR it doesn't count ;) ).

I worked with Leonard on this years ago.  The desire was to provide a little kick in the pants especially for high inertia vehicles to help get the acceleration going.  The delay in heli's is on the order of 100 ms and a little FF D could help reduce that.  Certainly you could do better with feedforward shaping using acceleration (which Leonard already has in the works :)).  However this is probably the next best thing.

just my 2 cents

andyp1per commented:
Just to add to @bnsgeyer 's comments:

1. Agreed, but in testing on copters I found that notching the target was pretty ineffectual for frame resonance

2. This is absolutely essential. I have set this up on a couple of copters now and you must filter the error - the target does not cut it at all. I can dig out more logs - but the logs I posted are representative. The key point is that this is not just frame resonance - it is a feedback effect coupled between the frame and the motors and this is why filtering the error is both effective and necessary. If it was pure resonance then, granted, target would be enough.

3. D_FF is the part I am most excited about - in highly aggressive flight this gives much more edge to maneuvers - and it makes it feel much more locked in. I will be flying this on all my quads from now on - having experienced it there is no way I am going back :) The D_FF change also happens to be extremely small relative to the rest. 

@IamPete1 apologies, I wasn't ignoring your comment about splitting the PR - happy to do so, but think questions of structure need addressing first - @lthall advised to keep it all together now so that it can be discussed effectively. It's also obviously a pain doing all these doc updates :-P

IamPete1 commented:
> @IamPete1 I would be interested to hear more about these. I am not familiar with what else could be used to do this but that isn't saying much ;)
> 

When reading up for https://github.com/ArduPilot/ardupilot/pull/23432 I found things like https://www.zaber.com/articles/input-shaping-for-vibration-reduction#A_Generic_ZV_Algorithm and https://www.extrica.com/article/16725. Input shaping is also quite common in the 3D printer world, they use methods like MZV and EI. There are a few papers that use notch as types of input shaping, but it seems like its not common. I think this stuff is only valid when applying to the target, and maybe it should be one level above the PID controllers. Notches are still the right thing for filtering the error.

lthall commented:
Lots of good discussion here. And a lot to cover too.

FFD

The missing FFD is the main cause of attitude control delay in multirotors. I have known this for a long time but have not included this because we had too much noise and I could not justify the tuning complexity. Now we have much less noise due to the development in notch filters and more and more people are using arducopter on small aircraft used in acrobatic applications.

The interesting thing about beta flight was they just called it ff (normal FF would be silly) and it wasn't until I saw a more detailed drawing of their controller that I realised it was actually FFD. FFD is an obvious improvement if done correctly so I was happy to see this put in place and shown to be valuable.

My intention is to use real time S-Curves to upgrade the input shaping for our attitude control and this will produce a noise free acceleration component that we can use to drive the FFD. Until then Andy's implementation is working very well on low noise aircraft.


Notch in PID

First thing to point out is I didn't implement a notch here, I used a Peaking Filter. This is why we carry around an attenuation as well. A Peaking Filter becomes a notch if you ask for infinite attenuation.

Peaking filters have been used for a long time to attenuate gain peaks in our plan transfer functions to achieve better tuning. They can also be used to increase gain at a frequency. All this gets into some pretty serious control applications. However, I was aware that we may need this feature when I first implemented the fixed notch. Working with Bill on Heli's I realised we actually needed separate notches on each axis so our fixed notch would not be good enough.

T vs E

We have two basic controller types. Rate dominated where FF is the main control path and Acceleration dominated where the PID is the main control path. Depending on the airframe you may require the T, E or both. Heli is a case where you will require both I think to get the best tune. I suspect that Bill did not see the benifit because he did not retune the PID to take advantage of the notch. I find that heli's are very sensitive the PID P term and it is very small due to this acceleration.

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.cpp:
```diff
@@ -74,6 +75,38 @@ const AP_Param::GroupInfo AC_AttitudeControl_Multi::var_info[] = {
 
+    // @Param: RAT_RLL_D_FF
+    // @DisplayName: Roll Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the error
```
are we missing the Range, Units and USer for RAT_RLL_D_FF?

andyp1per replied:
Thanks fixed as best I can

tridge commented:
we had a long discussion on todays dev call, conclusion was:
 - have this under an AC_PID_ADVANCED_ENABLED, off by default
 - initially only via custom build server
 - add an _ADV advanced parameter as an enable parameter that hides all these new parameters

IamPete1 commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -137,15 +210,42 @@ float AC_PID::update_all(float target, float measurement, float dt, bool limit,
+        _target_derivative = 0.0f;
+#if AC_PID_ADVANCED_ENABLED
+        if (_adv_enable) {
+            if (!is_zero(_notch_T_center_freq_hz.get())) {
```
Would it be cleaner to move this down in the notch is self?

andyp1per replied:
The parameters is in a different class - difficult to see how this is going to make things simpler

IamPete1 commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -64,12 +66,62 @@ const AP_Param::GroupInfo AC_PID::var_info[] = {
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS_DEFAULT_POINTER("D_FF", 33, AC_PID, _kdff, default_kdff),
+
+    // @Param: NTF
+    // @DisplayName: PID Target notch Filter center frequency
+    // @Description: PID Target notch Filter center frequency in Hz.
+    // @Range: 10 495
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("NTF", 34, AC_PID, _notch_T_center_freq_hz, 0),
+
+    // @Param: NEF
+    // @DisplayName: PID Error notch Filter center frequency
+    // @Description: PID Error notch Filter center frequency in Hz.
+    // @Range: 10 495
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("NEF", 35, AC_PID, _notch_E_center_freq_hz, 0),
+
+    // @Param: NBW
+    // @DisplayName: PID notch Filter bandwidth
+    // @Description: PID notch Filter bandwidth in Hz.
+    // @Range: 5 250
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("NBW", 36, AC_PID, _notch_bandwidth_hz, 0),
+
+    // @Param: NATT
+    // @DisplayName: PID notch Filter attenuation
+    // @Description: PID notch Filter attenuation in dB.
+    // @Range: 5 50
+    // @Units: dB
+    // @User: Advanced
+    AP_GROUPINFO("NATT", 37, AC_PID, _notch_attenuation_dB, 40),
```
I will just note that there are two center frequencies here but only one set of bandwidth and attenuation. This means you have to have them both the same, or just use target or error but not both.

andyp1per replied:
Yeah, noted. I originally had all params. Interested in @lthall and @bnsgeyer 's thoughts on this

bnsgeyer replied:
@andyp1per and @IamPete1   
Leonard and I decided to make users use the same parameters for target and error because it was unlikely that someone would need to have the notch for the target and error signal be different. And it was more beneficial to reduce the parameter count

IamPete1 commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -262,8 +375,12 @@ float AC_PID::get_d() const
-    return _target * _kff;
+    _pid_info.FF = _target * _kff
+#if AC_PID_ADVANCED_ENABLED
+        + _target_derivative * _kdff
```
Should we be logging this separately?

andyp1per replied:
Yeah seems like a good idea

andyp1per replied:
Actually we don't log the derivative so that would be inconsistent with that. Happy to go either way.

IamPete1 commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -120,6 +173,26 @@ void AC_PID::slew_limit(float smax)
+    if (!is_zero(_notch_T_center_freq_hz.get()) &&
+        (!is_equal(sample_rate, _target_notch.sample_freq_hz())
+         || !is_equal(_notch_T_center_freq_hz.get(), _target_notch.center_freq_hz()))) {
+        _target_notch.init(sample_rate, _notch_T_center_freq_hz, _notch_bandwidth_hz, _notch_attenuation_dB);
```
won't the init already not do anything if the center frequency is the same?

andyp1per replied:
No, it will get recalculated

andyp1per replied:
The notch filter does not actually record the sample rate, so it can't be done in the filter

IamPete1 replied:
There is a variable for it...
https://github.com/ArduPilot/ardupilot/blob/daf427f6d612614f08a1ad770c42109f989e8dc5/libraries/Filter/NotchFilter.h#L45

andyp1per commented:
Vehicle with no static notches - massive resonance:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/72217f6f-b346-4f7f-a439-6bfd9a3a5ed4)

Vehicle with static notch at 229/115 - mostly resolved:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/34c6070f-b3b7-428e-8b9c-c9c676bdd3ad)

Vehicle with PID error notch on roll only - better:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/b03195b5-4a9b-4023-a7e3-e967154d4626)

Vehicle with PID target notch on roll - massive resonance:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/a074b771-8694-4750-9afd-68a4928f0de3)

lthall commented:
This looks great.

I have been considering my choice of using F0 (centre frequency) and BW (bandwidth) and I think I made the wrong choice. I thought that bandwidth would be easier to understand for people but instead it means that everybody needs to set it with the centre frequency. This has lead to more setup problems and support issues than there needed to be.

My preference as an engineer was to use Q but I thought that people would find it harder to understand.

The Q or Quality factor is defined as the ratio of the notch or peak frequency F 0 and the bandwidth BW : Q = F 0 / BW . The quality factor is a measure of how well you can isolate the desired frequency from the other frequencies.

I would like to propose that we use Q instead of BW. This will make both notches more useable if they are set to different frequencies.

Further I would propose that we do the same for the harmonic notches on the gyro. This would be a very easy parameter conversion and would mean that we rarely need to change it.

andyp1per commented:
I'm ok with Q if everyone agrees. So the parameter would be ```_NQ```

IamPete1 commented:
Q might have been better from the start, but I don't think its worth changing it now. Lots of users are now familiar with bandwidth  and it is what all the tools and documentation use. 

I could add a calculated Q to the filter tools, so it would not be a input, but you could see what Q the current settings would have.

bnsgeyer commented:
I think Q makes a lot of sense as it sizes the bandwidth based off of the center freq. but I can see how this would be controversial since we have come so far with using bandwidth

andyp1per commented:
Changed to Q

tridge commented:
proportion of frequency (ie. 1/Q) would be easier to explain and would match what we've been telling users to use for years (ie. we tell them to use bandwidth = half of frequency)

tridge commented:
discussed on dev call, proposed we have:
 - FILT1_ENABLE, FILT2_ENABLE etc for 8 filters
 - ATC_RAT_RLL_FIDX for which filter to apply, defaults to 0 for no filter

andyp1per commented:
Ok, after some discussion with @tridge and @lthall here is what I propose:

```ATC_RAT_XXX_ADV``` will be deleted

two new parameters in the PIDs
```
ATC_RAT_XXX_NFT
ATC_RAT_XXX_NFE
```
For "Notch Filter Target" and "Notch Filter Error". I strongly disagree with trying to make this some kind of generic filter mechanism because (a) this is not a problem we have, we should not overdesign for an unknown future (b) the names can be changed in the future and (c) it will be way too confusing for users. I am already not convinced that the current proposal is making things easier for users, but I accept it gives us some more flexibility - but this should not be done to the detriment of usability.
These parameters will reference a notch filter configuration, the default "0" meaning off.
The notch filter configuration will take the following parameters:
```
NOTCH_FILTx_FREQ
NOTCH_FILTx_Q
NOTCH_FILTx_ATT
```
where x is 1-8 (seems too many to me - why so many?)
Please agree/disagree @bnsgeyer @lthall @tridge

timtuxworth commented:
> Ok, after some discussion with @tridge and @lthall here is what I propose:
> 
> `ATC_RAT_XXX_ADV` will be deleted
> 
> two new parameters in the PIDs
> 
> ```
> ATC_RAT_XXX_NFT
> ATC_RAT_XXX_NFE
> ```
> 
> For "Notch Filter Target" and "Notch Filter Error". I strongly disagree with trying to make this some kind of generic filter mechanism because (a) this is not a problem we have, we should not overdesign for an unknown future (b) the names can be changed in the future and (c) it will be way too confusing for users. I am already not convinced that the current proposal is making things easier for users, but I accept it gives us some more flexibility - but this should not be done to the detriment of usability. These parameters will reference a notch filter configuration, the default "0" meaning off. The notch filter configuration will take the following parameters:
> 
> ```
> NOTCH_FILTx_FREQ
> NOTCH_FILTx_Q
> NOTCH_FILTx_ATT
> ```
> 
> where x is 1-8 (seems too many to me - why so many?) Please agree/disagree @bnsgeyer @lthall @tridge

Comment from a "stupid people" :- this seems like this is similar/related to Harmonic Notch filters. What I'm finding challenging is the whole new standard for abbreviations for "notch" and "filter" from HNTCH and the unfortunately named HNTC2 - seems like notch is abbreviated "NTCH" when used in harmonic notches but not abbreviated and/or abbreviated to NOTCH/NF here. 

[Aside: What I've finally figured out with notch filters, is that the name is about the solution, whereas the problem seems to be motor/engine noise, so from an end user it would have probably cut about 6 months off my learning curve if they had been called "Noise filters". I never could figure out why I cared about "notches" but it's easy to understand why I care about vibration/noise.]

So for this one - again you are talking about notches - but what is the problem that it's solving? It seems to be "frame resonance", is that right? So this might help to create a more meaningful name that isn't overloading the existing use of "Notch" and "Filter" and just creating more confusion for us stupid people.

If this is  "Frame resonance notch filter" then ... the key part is frame resonance, and "notch" is just an implementation detail, so ...

How about FRAMR_FLTRx_FREQ/Q/ATT and 
ATC_RAT_XXX_FRT
ATC_RAT_XXX_FRE
and the equivalent QuadPlane parameters (I think)
Q_A_RAT_XXX_FR[TE]

lthall commented:
> where x is 1-8 (seems too many to me - why so many?)

8 is a good number. In just roll, pitch, yaw and thrust we have 8 potential filters. We won't use that much normally but I would expect this to expand over time to be used in a few other places too.

> I am already not convinced that the current proposal is making things easier for users, but I accept it gives us some more flexibility - but this should not be done to the detriment of usability.

Yep, it is a line ball at this time but I think it will come out to be a good decision if we grow this concept. Growing this concept is the main reason for suggesting a more generic filter name but as you point out it isn't important right now. Having 8 filters is another small encouragement for this to grow.

> So for this one - again you are talking about notches - but what is the problem that it's solving?

We are able to use these to solve multiple problems so naming based on the problem isn't sensible. We often have conversations about making it easier for new users to work out how to setup and tune aircraft.

The line in the sand that I draw is we should not make it difficult for someone that does understand the engineering, math and control theory by using non-sensical naming conventions to make it easier for someone who doesn't understand the engineering, math and control theory. Every time I have relaxed this stance (Q vs BW for example) I have found it has cost the project by some of the following:

- poorer outcomes, 
- less accurate parameter descriptions,
- less professional impressions, 
- more support overhead to both beginners and experts, 
- more complex setup procedures, 
- and more parameters.

Engineers are just as lazy as everybody else so there is a strong tendency for mature engineering naming conventions to be the simplest and lowest cost solution to the communication problem. Sometimes it just takes some time to understand why it is the best and simplest solution.

lthall commented:
ATC_RAT_XXX_FT
ATC_RAT_XXX_FE
FLTx_TYPE
FILTx_NOTCH_FREQ
FILTx_NOTCH_Q
FILTx_NOTCH_ATT

andyp1per commented:
OK proposal is now:

```
ATC_RAT_XXX_NFT
ATC_RAT_XXX_NFE
FILTx_TYPE
FILTx_NOTCH_FREQ
FILTx_NOTCH_Q
FILTx_NOTCH_ATT
```
FILTx_TYPE can only ever be 1 for a notch filter. We already have FLTE and FLTT so trying to a more generic FT and FE is highly confusing and pointless at this stage. If we were to ever make this more generic the names are trivial to change. Since most of the parameters are hidden inside FILTx we can remove the need for the _ADV setting.

timtuxworth commented:
> OK proposal is now:
> 
> ```
> ATC_RAT_XXX_NFT
> ATC_RAT_XXX_NFE
> FILTx_TYPE
> FILTx_NOTCH_FREQ
> FILTx_NOTCH_Q
> FILTx_NOTCH_ATT
> ```
> 
> FILTx_TYPE can only ever be 1 for a notch filter. If it is 0 then the filter is disabled and the parameters will disappear. We already have FLTE and FLTT so trying to a more generic FT and FE is highly confusing and pointless at this stage. If we were to ever make this more generic the names are trivial to change. Since most of the parameters are hidden inside FILTx we can remove the need for the _ADV setting.

Am I correct in assuming then that _currently_ we only have one filter type (harmonic notch) - but TYPE gives us the option to add others in the future. Is that the idea?

So is this how it will work? 

FILT1_TYPE = 1 - a notch filter with a notch at 45 Hz with 3 harmonics
FILT2_TYPE = 2 a low pass filter (parameters tbd) 
FILT3_TYPE = 3 a high pass filter (paramstaeters tbd)

Then by setting different filter definition values for the different sources, I could apply different filters to each like this?
ATC_RAT_ROL_NFT = 1
ATC_RAT_ROL_NFE = 2
ATC_RAT_PIT_NFT = 3
ATC_RAT_YAW_NFT = 3

[I'm throwing in low and high pass just as examples of different filter types for illustration. Not even suggesting I know enough to know if they even make sense as options.]

andyp1per commented:
@timtuxworth see the autotest for an example of usage

tridge commented on libraries/Filter/AP_Filter.cpp:
```diff
@@ -0,0 +1,152 @@
+            AP_Param::invalidate_count();
+        }
+    }
+    if (enable == 0 && count > 0) {
```
i think we can remove the enable parameter completely, reducing the number of steps to configure by 1

andyp1per replied:
Removed and added update

tridge reviewed (changes requested):
worth testing this with valgrind in sitl, remove enable parameter as discussed
possibly do a 1Hz update call to avoid the reboot on initial setup?

andyp1per commented:
valgrind is clean

andyp1per commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -70,12 +72,36 @@ const AP_Param::GroupInfo AC_PID::var_info[] = {
+
+#if AP_FILTER_ENABLED
+    // @Param: NTF
+    // @DisplayName: PID Target notch Filter
```
Fixed

andyp1per commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -70,12 +72,36 @@ const AP_Param::GroupInfo AC_PID::var_info[] = {
+    AP_GROUPINFO("NTF", 34, AC_PID, _notch_T_filter, 0),
+
+    // @Param: NEF
+    // @DisplayName: PID Error notch Filter
```
Fixed

andyp1per commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -275,6 +356,17 @@ void AC_PID::load_gains()
     _filt_T_hz.load();
     _filt_E_hz.load();
     _filt_D_hz.load();
+    _kdff.load();
```
I just removed them as agreed

IamPete1 commented:
It would be more runtime calculations, but we could add a filter method to take a variable time step. This would completely remove all the `set_notch_sample_rate` calls all over the place. Its no different than the loop rate update dynamic notch, we could save calculations by only updating the sample rate and not the target frequency.

IamPete1 commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -253,8 +334,8 @@ float AC_PID::get_d() const
 {
-    _pid_info.FF = _target * _kff;
-    return _target * _kff;
+    _pid_info.FF = _target * _kff + _target_derivative * _kdff;
```
This is a existing bug but this but of the calculation should be done in the main `update_all` function. This removes a race condition between when we update, when we log and when we get the ff.

andyp1per replied:
Fixed.

tridge commented:
@andyp1per would be nice to add PID FF to PID log

andyp1per commented:
> @andyp1per would be nice to add PID FF to PID log

I discussed this with @IamPete1 and @lthall - the feeling is that we are already logging FF - so it is being logged - Pete already has PID logging improvements planned and taking it out if its wrong is harder than adding it later.

andyp1per commented:
> It would be more runtime calculations, but we could add a filter method to take a variable time step. This would completely remove all the `set_notch_sample_rate` calls all over the place. Its no different than the loop rate update dynamic notch, we could save calculations by only updating the sample rate and not the target frequency.

I actually was going to do this originally but the filter maths do not allow it according to @lthall

tridge commented:
@andyp1per just did a flight on Nessie (current name for @MichelleRos 5" quad). Good news is it survived!
It sure would be nice to have the D_FF in the PID logs.
The log is here:
http://uav.tridgell.net/tmp/00000056.BIN
can you tell how much the D_FF was impacting flight?
There are some very bad pitch events when throttle is lowered quickly. Should a higher D_FF help?

andyp1per commented:
> @andyp1per just did a flight on Nessie (current name for @MichelleRos 5" quad). Good news is it survived! It sure would be nice to have the D_FF in the PID logs. The log is here: http://uav.tridgell.net/tmp/00000056.BIN can you tell how much the D_FF was impacting flight? There are some very bad pitch events when throttle is lowered quickly. Should a higher D_FF help?

It's pretty easy to see what D_FF is doing because in copter there is no FF component so it is all coming from D_FF:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/e39509f4-0114-40a3-93a9-dd46b5744697)

You can see that the effect is not huge here and the pitch events are coming directly from P. I will do a bit more digging, because this seems both very weird and bad!

andyp1per commented:
The pitch events are coming through the target, but the target and actual are going in opposite directions - I will need to talk to @lthall about what is going on here>

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/e3eac145-4534-4795-9fee-3244a10f2925)

andyp1per commented:
@tridge @MichelleRos you have ATC_RAT_PIT_FLTE set to 40. Why is this? If I had to guess I would say this is the cause of the problem but will need to talk to Leonard.

MichelleRos commented:
@andyp1per Looking at the outputs in that log, it seems like the PID was trying to turn the copter in the right direction (back two motors much higher than the front two) but it was still pitching up, which seems to just be that pitch nod issue that needs higher PID gains on rapid throttle changes...
![image](https://github.com/ArduPilot/ardupilot/assets/64898873/c444db6d-854b-4ff0-a350-e26900e4db84)

I've had that ATC_RAT_PIT_FLTE = 40 for quite a while and it hasn't caused issues before. I normally see a smallish pitch bob on rapid throttle up. But I'll note that when we added D_FF, we ended up having to reduce some of the other gains to stop the oscillation, which I think is more likely to be what made it so bad.

Here's the log for the next flight, where we went back to (mostly) the original gains and then made changes one by one during the acro flight: https://drive.google.com/file/d/1b0u_BE-2u6vtoSUBUrEzXrhFLhhV7GOU/view?usp=sharing

lthall commented:
This has nothing to do with the DFF. Michelle is correct, this is the pitch nod issue. The standard things that drive this is:

1. Offset CG
2. Flexible props
3. Flexible arms
4. damaged prop
5. mismatched motor kv

There are a few others too.

andyp1per commented:
I added logging and retested and found that PID notches had been broken by the AP_Filter updates. I have now fixed that, retested and re-flown and now all looks good.

MichelleRos commented:
@lthall Thanks. The frame is CF so no flex, props are the standard plastic ones for 5" quads and still in very good condition. The four motors are the exact same brand & model. So it's only offset CG that can be it, but I've got that as close as I can, so the rest is for the controllers to deal with. It's pretty good already if you look at the second log, so I think it just needs some fine-tuning.

@andyp1per I think I should try out/tune your "anti-gravity" code to deal with the nodding - I just realised that I haven't yet tried it out.

I just found it a bit interesting that we had to lower the other gains when we added D_FF. In that case, I wonder where the balance is between them, since lowering the other gains too much results in worse control...

andyp1per commented:
> I just found it a bit interesting that we had to lower the other gains when we added D_FF. In that case, I wonder where the balance is between them, since lowering the other gains too much results in worse control...

Lowering the other gains is the problem. You should not need to do this. Note that you will get noise coming through the target with D_FF, so if you have a noisy build you might need to decrease FLTT. Tridge said that you could hear oscillation even without D_FF - so I think you should try and fix that first. It is possible to measure D_FF, what we need is a log with sharp twitches on the axes that you want to add D_FF to. So for example do sharp stick inputs right to left and back again.

MichelleRos commented:
@andyp1per We had to because there were stutters/oscillation as soon as we added D_FF. Without D_FF there wasn't much oscillation - it was on the edge but still ok. Then D_FF put it over the edge and we had to lower gains.
I highly doubt this is a noisy build at all - the vibration is crazy low, and we've been able to raise the gains much higher and it flies much better than Braveheart did. 

And having to lower any of the lowpass filters would once again reduce performance by increasing lag, wouldn't it? So then it's again the question of does it perform better with D_FF but lower filters, or is it better to not use D_FF and be able to use higher gains and filters...

andyp1per commented:
On all my builds I have been able to add D_FF without seeing the problems you allude to. My flights on the weekend were absolutely superb - the best control I have ever had. So something is off between your builds and by builds. Perhaps the thing to do is try and calculate what your D_FF should be and then work from there.

MichelleRos commented:
Well, I'm not really keen on retuning my quad from scratch just because it's "supposed" to give better control. My quad already flies pretty good, so I don't see very good motivation to spend another few hours retuning to try to use this when I could just enjoy flying it instead.
Maybe I'll try adding a bit of D_FF in again at some point, but that's about all... I think increasing the various acceleration limits etc is more where improvements in this quad's control will come from.

andyp1per commented:
@MichelleRos pretty sure this is your problem:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/c3b8dab1-802f-4917-85e4-460fd05849a1)

andyp1per commented:
Interestingly if it is resonance its being folded down:

![image](https://github.com/ArduPilot/ardupilot/assets/2893260/f30ec34a-f1b4-4c07-9b19-4abf9daf8237)

You might benefit from using a loop rate of 800

MichelleRos commented:
How is the Hntch opts the problem? That's Multi-Source, Update at loop rate and Triple notch. Isn't that exactly what I'm meant to have?
And letting the notch go down further caused too much phase lag which caused oscillation, that's why it has to be reasonably high.
In any case, lowering the mot_thst_expo to 0.7 actually fixed the low throttle oscillation, so looks like it was simply that this one's expo is a bit lower than Braveheart's was.
And I've tried setting the looprate to 800 before (with Braveheart) and I got a bunch of long loops. You can say it's just my fault again, but I haven't yet seen a log from a copter with the looprate set to 800 (and all of the filtering etc enabled) without also seeing long loops.

andyp1per commented:
Its not the options that is the problem - its the narrowness of the notch. At best on super clean aircraft using the triple notch F/6 is the absolutely limit I have found for bandwidth. At F/12 you are going to be missing noise - and you can see that in the FFT.

tridge commented:
@andyp1per this PR against your PR adds fixed wing support for D_FF:
https://github.com/andyp1per/ardupilot/pull/30

tridge commented:
@priseborough can I get you to look at the scaling of the D_FF on fixed wing? Should we be scaling the same as FF?

priseborough commented:
The running a feed forward gain on the derivative of demanded rate is effectively a gain from angular acceleration demand to control surface. The angular acceleration produced by an given control surface deflection is proportional to dynamic pressure. This means the D_FF term should be scaled using scaling speed^2, not scaling speed^1 as per the proportional FF gain.

tridge commented:
> This means the D_FF term should be scaled using EAS^2, not TAS^1 which the proportional FF term uses

thanks @priseborough - it turns out that is what I implemented, so the PR is correct (I think this was just luck, not proper understanding)

IamPete1 commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -253,45 +336,40 @@ float AC_PID::get_d() const
-    _kimax.load();
-    _kimax.set(fabsf(_kimax));
-    _kpdmax.load();
-    _kpdmax.set(fabsf(_kpdmax));
```
Presumable this is because autotune does not change I max and PD max?

andyp1per replied:
Yes

IamPete1 commented:
I have dismissed my stale review, so this PR can be merged.

I'm not convinced this is a great way for users to configure filters. Its seems overly confusing. But I cannot think of any better way with our current parameter structure. If we move the parameters down to a more logical place were left with very few characters left for names.  

The functionality will be very beneficial.

tridge commented on AntennaTracker/Parameters.cpp:
```diff
@@ -387,6 +387,25 @@ const AP_Param::Info Tracker::var_info[] = {
+    // @Param: PITCH2SRV_D_FF
+    // @DisplayName: Pitch Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.1
```
range should start with 0, or default is out of range

andyp1per replied:
Fixed

tridge commented on AntennaTracker/Parameters.cpp:
```diff
@@ -464,6 +483,25 @@ const AP_Param::Info Tracker::var_info[] = {
+    // @Param: YAW2SRV_D_FF
+    // @DisplayName: Yaw Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.1
```
range start at zero

andyp1per replied:
Fixed

tridge commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -357,6 +357,16 @@ void Plane::one_second_loop()
+    const float loop_rate = AP::scheduler().get_filtered_loop_rate_hz();
+#if HAL_QUADPLANE_ENABLED
+    if (quadplane.available()) {
+        quadplane.attitude_control->set_notch_sample_rate(loop_rate);
```
when these notch rates change, will that inject a disturbance in the notch filter? should we only change when disarmed?

andyp1per replied:
This is a poor-mans equivalent to the delta_t stuff which cannot be used for notches. Leonard was ok with the behavior and my testing has not shown any anomalous behavior

tridge commented on Blimp/Parameters.cpp:
```diff
@@ -729,6 +729,26 @@ const AP_Param::Info Blimp::var_info[] = {
+    // @Param: POSYAW_D_FF
+    // @DisplayName: Position (yaw) Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.1
```
range start at zero

andyp1per replied:
Fixed

tridge commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -126,6 +153,37 @@ void AC_PID::slew_limit(float smax)
+        }
+        AP_Filter* filter = AP::filters().get_filter(_notch_T_filter);
+        if (filter != nullptr && !filter->setup_notch_filter(*_target_notch, sample_rate)) {
+            delete _target_notch;
```
we need to set _target_notch to nullptr here or we could use deleted memory

andyp1per replied:
Fixed

tridge commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -126,6 +153,37 @@ void AC_PID::slew_limit(float smax)
+        }
+        AP_Filter* filter = AP::filters().get_filter(_notch_E_filter);
+        if (filter != nullptr && !filter->setup_notch_filter(*_error_notch, sample_rate)) {
+            delete _error_notch;
```
need to set _error_notch to nullptr

andyp1per replied:
Fixed

tridge commented on libraries/APM_Control/AP_PitchController.cpp:
```diff
@@ -136,6 +136,24 @@ const AP_Param::GroupInfo AP_PitchController::var_info[] = {
+    // @Param: _RATE_D_FF
+    // @DisplayName: Pitch Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.03
```
needs zero start range

andyp1per replied:
Fixed

tridge commented on libraries/Filter/AP_Filter.cpp:
```diff
@@ -0,0 +1,155 @@
+        return;
+    }
+
+    // make sure all custom rotations are allocated
```
custom rotations?

andyp1per replied:
Fixed

andyp1per commented on libraries/APM_Control/AP_RollController.cpp:
```diff
@@ -119,6 +119,24 @@ const AP_Param::GroupInfo AP_RollController::var_info[] = {
+    // @Param: _RATE_D_FF
+    // @DisplayName: Roll Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.03
```
Fixed

andyp1per commented on libraries/APM_Control/AP_YawController.cpp:
```diff
@@ -150,6 +150,24 @@ const AP_Param::GroupInfo AP_YawController::var_info[] = {
+    // @Param: _RATE_D_FF
+    // @DisplayName: Yaw Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.03
```
Fixed

andyp1per commented on libraries/APM_Control/AR_AttitudeControl.cpp:
```diff
@@ -146,6 +146,24 @@ const AP_Param::GroupInfo AR_AttitudeControl::var_info[] = {
+    // @Param: _STR_RAT_D_FF
+    // @DisplayName: Steering control Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.03
```
Fixed

andyp1per commented on libraries/APM_Control/AR_AttitudeControl.cpp:
```diff
@@ -229,6 +247,24 @@ const AP_Param::GroupInfo AR_AttitudeControl::var_info[] = {
+    // @Param: _SPEED_D_FF
+    // @DisplayName: Speed control Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.03
```
Fixed

andyp1per commented on libraries/APM_Control/AR_AttitudeControl.cpp:
```diff
@@ -372,6 +408,24 @@ const AP_Param::GroupInfo AR_AttitudeControl::var_info[] = {
+    // @Param: _BAL_D_FF
+    // @DisplayName: Pitch control Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.03
```
Fixed

andyp1per commented on libraries/APM_Control/AR_AttitudeControl.cpp:
```diff
@@ -463,6 +517,24 @@ const AP_Param::GroupInfo AR_AttitudeControl::var_info[] = {
+    // @Param: _SAIL_D_FF
+    // @DisplayName: Sail Heel Derivative FeedForward Gain
+    // @Description: FF D Gain which produces an output that is proportional to the rate of change of the target
+    // @Range: 0.001 0.03
```
Fixed
